windows/src/desktop/kmshell/main/UImportOlderVersionKeyboards9Plus.pas

```diff
@@ -0,0 +1,383 @@
+    procedure UnregisterKMTIPFromOldVersion;
+
+    property SRegKey_KeymanEngineOld_InstalledPackages_CU: string read FSRegKey_KeymanEngineOld_InstalledPackages_CU write FSRegKey_KeymanEngineOld_InstalledPackages_CU;
+    property SRegKey_KeymanEngineOld_InstalledKeyboards_CU: string read FSRegKey_KeymanEngineOld_InstalledKeyboards_CU write FSRegKey_KeymanEngineOld_InstalledKeyboards_CU;
+    property SRegKey_KeymanEngineOld_ActiveKeyboards_CU: string read FSRegKey_KeymanEngineOld_ActiveKeyboards_CU write FSRegKey_KeymanEngineOld_ActiveKeyboards_CU;
+    property c_clsidKMTipTextService_Old: TGUID read Fc_clsidKMTipTextService_Old write Fc_clsidKMTipTextService_Old;
+  end;
+
+implementation
+
+uses
+  Vcl.Graphics,
+  System.Win.ComObj,
+  Winapi.ActiveX,
+  input_installlayoutortip,
+  keyman_msctf,
+  utilexecute,
+  utilolepicture,
+  GetOsVersion,
+  Glossary,
+  KeymanVersion,
+  kmint,
+  Windows,
+  Classes,
+  ErrorControlledRegistry,
+  kmxfile,
+  KeymanPaths,
+  Keyman.System.UpgradeRegistryKeys,
+  RegistryKeys,
+  StrUtils,
+  SysUtils,
+  utildir,
+  utilsystem,
+  utiltsf,
+  UImportOlderKeyboardUtils;
+
+
+function ConvertLangIDToLocale(LangID: Integer; var Locale: string): Boolean;
+var
+  buf: array[0..LOCALE_NAME_MAX_LENGTH-1] of char;
+begin
+  Result := LCIDToLocaleName(LangID, buf, LOCALE_NAME_MAX_LENGTH, 0) > 0;
+  if Result
+    then Locale := buf
+    else Locale := '';
+end;
+
+function GetDefaultHKL: HKL;   // I3581   // I3619   // I3619
+begin
+  if not SystemParametersInfo(SPI_GETDEFAULTINPUTLANG, 0, @Result, 0) then
+    Result := 0;
+end;
+
+procedure TImportKeyman9PlusKeyboard.Import;
+var
+  keysWrite, keys: TStringList;
+  i: Integer;
+  n: Integer;
+  s: string;
+begin
+  OleCheck(CoCreateInstance(CLASS_TF_InputProcessorProfiles, nil, CLSCTX_INPROC_SERVER,
+                          IID_ITfInputProcessorProfiles, pInputProcessorProfiles));
+
+  if not Supports(pInputProcessorProfiles, IID_ITfInputProcessorProfileMgr, pInputProcessorProfileMgr) then   // I3743
+  begin
+    LogFailure('Missing interface IID_ITfInputProcessorProfileMgr');
+    Exit;
+  end;
+
+  values := TStringList.Create;
+  keys := TStringList.Create;
+  keysWrite := TStringList.Create;
+  regRead := TRegistryErrorControlled.Create;  // I2890
+  regWrite := TRegistryErrorControlled.Create;  // I2890
+  try
+    if FAdmin then  // I2361
+    begin
+      regRead.RootKey := HKEY_LOCAL_MACHINE;
+      // We use LM + CU here because we are sharing the LM and CU base paths in the backup
+      rootRead := SRegKey_UpgradeBackupPath_LM + FSRegKey_KeymanEngineOld_InstalledPackages_CU;  // I2642
+    end
+    else
+    begin
+      regRead.RootKey := HKEY_CURRENT_USER;
+      rootRead := SRegKey_UpgradeBackupPath_CU + FSRegKey_KeymanEngineOld_InstalledPackages_CU;  // I2642
+    end;
+
+    regWrite.RootKey := HKEY_LOCAL_MACHINE;
+    rootWrite := '\'+SRegKey_InstalledPackages_LM; // '\Software\...\Keyman Engine\x.x\Installed Packages';
```

Comment:
Is the commented out part still needed?

Reply:
also for ll 129, 148

Reply:
The comment was just to help the casual code reader know what the registry key is. What's the issue with it?